src/runtime/importhook.cs

```diff
@@ -141,189 +128,177 @@ internal static void SaveRuntimeData(RuntimeDataStorage storage)
-            if (res != IntPtr.Zero)
-            {
-                // There was no error.
-                if (fromlist && IsLoadAll(fromList))
+                if (Runtime.PyDict_SetItemString(root.dict, availableNsKey, newset.DangerousGetAddress()) != 0)
                 {
-                    var mod = ManagedType.GetManagedObject(res) as ModuleObject;
-                    mod?.LoadNames();
+                    throw new PythonException();
                 }
-                return res;
             }
-            // There was an error
-            if (!Exceptions.ExceptionMatches(Exceptions.ImportError))
+            finally
             {
-                // and it was NOT an ImportError; bail out here.
-                return IntPtr.Zero;
+                newset.Dispose();
             }
 
-            if (mod_name == string.Empty)
-            {
-                // Most likely a missing relative import.
-                // For example site-packages\bs4\builder\__init__.py uses it to check if a package exists:
-                //     from . import _html5lib
-                // We don't support them anyway
-                return IntPtr.Zero;
-            }
-            // Save the exception
-            var originalException = new PythonException();
-            // Otherwise,  just clear the it.
-            Exceptions.Clear();
-
-            string[] names = realname.Split('.');
+            AssemblyManager.namespaceAdded += OnNamespaceAdded;
+            PythonEngine.AddShutdownHandler(() => AssemblyManager.namespaceAdded -= OnNamespaceAdded);
+        }
 
-            // See if sys.modules for this interpreter already has the
-            // requested module. If so, just return the existing module.
-            BorrowedReference modules = Runtime.PyImport_GetModuleDict();
-            BorrowedReference module = Runtime.PyDict_GetItem(modules, py_mod_name);
+        /// <summary>
+        /// Removes the set of available namespaces from the clr module and 
+        /// removes the callback on the OnNamespaceAdded event.
+        /// </summary>
+        static void TeardownNameSpaceTracking()
+        {
+            AssemblyManager.namespaceAdded -= OnNamespaceAdded;
+            // If the C# runtime isn't loaded, then there are no namespaces available
+            Runtime.PyDict_SetItemString(root.dict, availableNsKey, Runtime.PyNone);
+        }
 
-            if (module != null)
+        static void OnNamespaceAdded(string name)
+        {
+            using (Py.GIL())
             {
-                if (fromlist)
+                var pyNs = Runtime.PyString_FromString(name);
+                try
                 {
-                    if (IsLoadAll(fromList))
+                    var nsSet = Runtime.PyDict_GetItemString(new BorrowedReference(root.dict), availableNsKey);
+                    if (!nsSet.IsNull || nsSet.DangerousGetAddress() != Runtime.PyNone)
```

Comment:
Should BorrowedReference have an IsNullOrNone property? It seems likely to be a common idiom.

Reply:
I'd keep separate `IsNull` and `IsNone`. Thought there's already `IsNone`.